File: shared/agent-core/src/monitors.ts
import type { MonitorDefinition, MonitorState } from "./types";

export class MonitorSystem {
  private monitors: MonitorDefinition[];
  private state: MonitorState;
  private timers: Map<string, ReturnType<typeof setInterval>> = new Map();
  private fireEvent: (eventName: string, context?: string) => Promise<void>;

  constructor(
    monitors: MonitorDefinition[],
    fireEvent: (eventName: string, context?: string) => Promise<void>
  ) {
    this.monitors = monitors;
    this.fireEvent = fireEvent;
    this.state = { failureCounts: {}, lastCheckAt: null, firedEvents: {} };
  }

  start(): void {
    for (const monitor of this.monitors) {
      if (!monitor.enabled) continue;

      // Run first check after 10 seconds
      setTimeout(() => this.checkMonitor(monitor), 10_000);

      const timer = setInterval(() => this.checkMonitor(monitor), monitor.intervalMs);
      this.timers.set(monitor.id, timer);
      console.log(`[monitor] Started: ${monitor.name} (every ${monitor.intervalMs / 1000}s)`);
    }
  }

  stop(): void {
    for (const [id, timer] of this.timers) {
      clearInterval(timer);
    }
    this.timers.clear();
    console.log("[monitor] All monitors stopped");
  }

  private async checkMonitor(monitor: MonitorDefinition): Promise<void> {
    if (monitor.config.type === "health_check") {
      await this.checkHealthEndpoints(monitor);
    }
    this.state.lastCheckAt = new Date().toISOString();
  }

  private async checkHealthEndpoints(monitor: MonitorDefinition): Promise<void> {
    const config = monitor.config;

    for (const endpoint of config.endpoints) {
      const key = `${monitor.id}:${endpoint.name}`;
      try {
        const controller = new AbortController();
        const timeout = setTimeout(() => controller.abort(), endpoint.timeoutMs || 5000);

        const resp = await fetch(endpoint.url, { signal: controller.signal });
        clearTimeout(timeout);

        if (resp.ok) {
          // Service recovered
          const wasFired = this.state.firedEvents[key];
          this.state.failureCounts[key] = 0;

          if (wasFired) {
            delete this.state.firedEvents[key];
            console.log(`[monitor] ${endpoint.name} recovered`);
            await this.fireEvent("server_recovered", `${endpoint.name} 서비스가 복구되었습니다.`);
          }
        } else {
          this.state.failureCounts[key] = (this.state.failureCounts[key] || 0) + 1;
        }
      } catch {
        this.state.failureCounts[key] = (this.state.failureCounts[key] || 0) + 1;
      }

      // Check threshold
      const failures = this.state.failureCounts[key] || 0;
      if (failures >= config.failureThreshold && !this.state.firedEvents[key]) {
        this.state.firedEvents[key] = new Date().toISOString();
        console.log(`[monitor] ${endpoint.name} DOWN (${failures} consecutive failures)`);
        await this.fireEvent(
          monitor.eventName,
          `${endpoint.name} 서비스가 ${failures}회 연속 응답 실패했습니다.\nURL: ${endpoint.url}\nProject: ${endpoint.project || "UNKNOWN"}`
        );
      }
    }
  }

  setEnabled(monitorId: string, enabled: boolean): void {
    const monitor = this.monitors.find((m) => m.id === monitorId);
    if (!monitor) return;
    monitor.enabled = enabled;

    if (!enabled) {
      const timer = this.timers.get(monitorId);
      if (timer) { clearInterval(timer); this.timers.delete(monitorId); }
    } else if (!this.timers.has(monitorId)) {
      const timer = setInterval(() => this.checkMonitor(monitor), monitor.intervalMs);
      this.timers.set(monitorId, timer);
    }
  }

  getStatus(): { monitors: Array<{ id: string; name: string; enabled: boolean; lastCheck: string | null; failures: Record<string, number> }> } {
    return {
      monitors: this.monitors.map((m) => {
        const failures: Record<string, number> = {};
        for (const ep of m.config.endpoints) {
          const key = `${m.id}:${ep.name}`;
          const count = this.state.failureCounts[key] || 0;
          if (count > 0) failures[ep.name] = count;
        }
        return { id: m.id, name: m.name, enabled: m.enabled, lastCheck: this.state.lastCheckAt, failures };
      }),
    };
  }

  isAllHealthy(): boolean {
    return this.monitors.every((m) =>
      m.config.endpoints.every((ep) => (this.state.failureCounts[`${m.id}:${ep.name}`] || 0) === 0)
    );
  }

  getHealthyCount(): { healthy: number; total: number } {
    let healthy = 0;
    let total = 0;
    for (const m of this.monitors) {
      for (const ep of m.config.endpoints) {
        total++;
        if ((this.state.failureCounts[`${m.id}:${ep.name}`] || 0) === 0) healthy++;
      }
    }
    return { healthy, total };
  }
}

File: content-pipeline/backend/src/agent/monitors.py
"""MonitorSystem — asyncio health checks + event firing. Port of agent-core/src/monitors.ts"""

import asyncio
import logging
from typing import Callable, Awaitable

import httpx

from src.agent.types import MonitorDefinition

logger = logging.getLogger(__name__)

FireEventFn = Callable[[str, str | None], Awaitable[None]]


class MonitorSystem:
    def __init__(
        self,
        monitors: list[MonitorDefinition],
        fire_event: FireEventFn,
    ):
        self.monitors = monitors
        self.fire_event = fire_event
        self.failure_counts: dict[str, int] = {}
        self.last_check_at: str | None = None
        self.fired_events: dict[str, str] = {}
        self._tasks: list[asyncio.Task] = []
        self._stopped = False

    def start(self):
        self._stopped = False
        for monitor in self.monitors:
            if not monitor["enabled"]:
                continue
            task = asyncio.create_task(self._monitor_loop(monitor))
            self._tasks.append(task)
            logger.info(f"[monitor] Started: {monitor['name']} (every {monitor['interval_sec']}s)")

    def stop(self):
        self._stopped = True
        for task in self._tasks:
            task.cancel()
        self._tasks.clear()
        logger.info("[monitor] All monitors stopped")

    async def _monitor_loop(self, monitor: MonitorDefinition):
        # First check after 10 seconds
        await asyncio.sleep(10)
        while not self._stopped:
            try:
                await self._check_monitor(monitor)
            except Exception as e:
                logger.error(f"[monitor] Check error for {monitor['name']}: {e}")
            await asyncio.sleep(monitor["interval_sec"])

    async def _check_monitor(self, monitor: MonitorDefinition):
        from datetime import datetime
        self.last_check_at = datetime.now().isoformat()

        async with httpx.AsyncClient() as client:
            for endpoint in monitor["endpoints"]:
                key = f"{monitor['id']}:{endpoint['name']}"
                timeout_ms = endpoint.get("timeout_ms", 5000)

                try:
                    resp = await client.get(
                        endpoint["url"],
                        timeout=timeout_ms / 1000,
                    )
                    if resp.is_success:
                        was_fired = key in self.fired_events
                        self.failure_counts[key] = 0
                        if was_fired:
                            del self.fired_events[key]
                            logger.info(f"[monitor] {endpoint['name']} recovered")
                            await self.fire_event(
                                "server_recovered",
                                f"{endpoint['name']} 서비스가 복구되었습니다.",
                            )
                    else:
                        self.failure_counts[key] = self.failure_counts.get(key, 0) + 1
                except Exception:
                    self.failure_counts[key] = self.failure_counts.get(key, 0) + 1

                failures = self.failure_counts.get(key, 0)
                if failures >= monitor["failure_threshold"] and key not in self.fired_events:
                    self.fired_events[key] = datetime.now().isoformat()
                    logger.warning(f"[monitor] {endpoint['name']} DOWN ({failures} consecutive failures)")
                    await self.fire_event(
                        monitor["event_name"],
                        f"{endpoint['name']} 서비스가 {failures}회 연속 응답 실패했습니다.\n"
                        f"URL: {endpoint['url']}\n"
                        f"Project: {endpoint.get('project', 'UNKNOWN')}",
                    )

    def get_status(self) -> dict:
        monitors = []
        for m in self.monitors:
            failures: dict[str, int] = {}
            for ep in m["endpoints"]:
                key = f"{m['id']}:{ep['name']}"
                count = self.failure_counts.get(key, 0)
                if count > 0:
                    failures[ep["name"]] = count
            monitors.append({
                "id": m["id"],
                "name": m["name"],
                "enabled": m["enabled"],
                "lastCheck": self.last_check_at,
                "failures": failures,
            })
        return {"monitors": monitors}

    def get_healthy_count(self) -> tuple[int, int]:
        healthy = 0
        total = 0
        for m in self.monitors:
            for ep in m["endpoints"]:
                total += 1
                key = f"{m['id']}:{ep['name']}"
                if self.failure_counts.get(key, 0) == 0:
                    healthy += 1
        return healthy, total
